Source/CarthageKit/Project.swift

```diff
@@ -587,21 +587,65 @@ public final class Project { // swiftlint:disable:this type_body_length
 		pinnedVersion: PinnedVersion,
 		toolchain: String?
 	) -> SignalProducer<URL, CarthageError> {
+		
+		typealias FrameworkUrlAndDestinationUrl = (frameworkUrlInUnarchivedFolder: URL, frameworkUrlInBuildForlder: URL)
+
 		return SignalProducer<URL, CarthageError>(value: zipFile)
 			.flatMap(.concat, unarchive(archive:))
 			.flatMap(.concat) { directoryURL -> SignalProducer<URL, CarthageError> in
+				// For all frameworks in the directory where the archive has been expanded
 				return frameworksInDirectory(directoryURL)
-					.flatMap(.merge) { url -> SignalProducer<URL, CarthageError> in
-						return checkFrameworkCompatibility(url, usingToolchain: toolchain)
+					.collect()
+					// Check if multiple framewroks resolve to the same unique destination URL in the Carhage/Build/ folder.
+					// This is needed because frameworks might overwrite each others.
+					.flatMap(.merge) { frameworksUrls -> SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError> in
+						return SignalProducer<URL, CarthageError>(frameworksUrls)
+							.flatMap(.merge) { url -> SignalProducer<URL, CarthageError> in
+								return platformForFramework(url)
+									.attemptMap { platform -> Result<URL, CarthageError> in
+										guard let destinationURLInWorkingDir = platform
+											.relativeURL?
+											.appendingPathComponent(url.lastPathComponent, isDirectory: true) else {
+												return .failure(.internalError(description: "failed to construct framework destination url from \(platform) and \(url.lastPathComponent)"))
+										}
+
+										return .success(self
+											.directoryURL
+											.appendingPathComponent(destinationURLInWorkingDir.path, isDirectory: true))
+								}
+							}
+							.uniqueValues()
+							.collect()
+							.flatMap(.merge) { uniqueDestinationUrls -> SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError> in
+								guard uniqueDestinationUrls.count == frameworksUrls.count else {
+									return SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError>(error: CarthageError.invalidArchive(description: "found multiple frameworks with the same unarchiving destination"))
+								}
+
+								let frameworkUrlAndDestinationUrlPairs = Array(zip(frameworksUrls, uniqueDestinationUrls)
+									.map { FrameworkUrlAndDestinationUrl(frameworkUrlInUnarchivedFolder:$0, frameworkUrlInBuildForlder: $1) })
+								return SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError>(frameworkUrlAndDestinationUrlPairs)
+							}
+					}
+					// Check if the framework are compatible with the current Swift version
+					.flatMap(.merge) { pair -> SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError> in
+						return checkFrameworkCompatibility(pair.frameworkUrlInUnarchivedFolder, usingToolchain: toolchain)
```

Comment:
Should we do this filter before the above error? Then you could bundle multiple Swift versions? Or maybe this is unimportant with ABI stability.

Reply:
If we picked only the swift version of the framework base on a smarter compatibility check then yes, the compatibility check should happen before. I can move it now, however the compatibility takes longer than checking for duplicate destination paths